[YouTubeBridge] download videos and attach them to the feed

sabberworm opened the pull request:
This PR adds a param `attach_format` to the YouTube bridge. When used, the bridge tries to download the video in the requested format and, if successful, attaches it to the feed. The feed should then be consumable as video or audio podcasts by most podcast players.

I made some minor changes in preparation: enclosure metadata like length and mime type can now be passed directly from a bridge instead of having to be extracted by the format. This changes the output of the plaintext feed somewhat for enclosures but otherwise shouldn’t have much of an impact (I assume the plaintext feed is only used for debugging and its format doesn’t need to stay backwards-compatible, right?). I also created a `NullCache` caching strategy which I found useful during develop.

I tried to adhere to the [Coding style policy](https://github.com/RSS-Bridge/rss-bridge/wiki/Coding-style-policy) but I couldn’t figure out how to name variables. The pattern I could discern was that function parameters and variables holding request parameters use snake_case and other variables use camelCase. If that’s incorrect, I can fix it.

I will add some more comments and questions inline.

sabberworm committed:
refactor(enclosure): allow setting mime type directly

sabberworm committed:
feat(cache): support disabling the cache by specifying null

sabberworm committed:
feat(docker): clean apt cache after installs

sabberworm committed:
feat(filecache): make getPath publicly usable

sabberworm committed:
feat(YouTube): allow video downloads to be served with feeds

sabberworm committed:
feat(YouTube): install ffmpeg and youtube-dl in docker container

sabberworm commented on Dockerfile:
```diff
@@ -3,7 +3,9 @@ FROM php:7-apache
-	&& apt-get --yes update && apt-get --yes install libxml2-dev zlib1g-dev libmemcached-dev \
+	&& apt-get --yes update \
+	&& apt-get --yes install libxml2-dev zlib1g-dev libmemcached-dev python3-pip ffmpeg \
+	&& apt-get clean \
```
I added clean as per the Dockerfile best practices.

sabberworm commented on bridges/YoutubeBridge.php:
```diff
@@ -58,10 +58,44 @@ class YoutubeBridge extends BridgeAbstract {
 				'exampleValue' => 10
-			)
+			),
+			'attach_format' => array(
```
I would like to have tested whether the `youtube-dl` command is installed before showing these options in the form but there are no computations allowed when initializing `const`s.
I thought I could simply move the parameters from the `PARAMETERS` to a method by overriding `getParameters()` from `BridgeAbstract` but it turns out `BridgeAbstract` uses `PARAMETERS` in a few locations directly instead of going via `getParameters()`. IMHO this should be refactored.

sabberworm replied:
Done

sabberworm commented on bridges/YoutubeBridge.php:
```diff
@@ -58,10 +58,44 @@ class YoutubeBridge extends BridgeAbstract {
 		)
 	);
 
+	const FILE_TYPES = array(
```
Similar to my comment about `PARAMETERS` above: I would have liked to have only one list of all the available formats so it’s impossible to update only one. But that would have meant assembling the params definition in code, which is not possible with `const`.

sabberworm replied:
Done

sabberworm commented on bridges/YoutubeBridge.php:
```diff
@@ -102,9 +136,105 @@ private function ytBridgeAddItem($vid, $title, $author, $desc, $time){
+			return;
+		}
+
+		$urlPrefix = (isset($_SERVER['HTTPS']) && $_SERVER['HTTPS'] == 'on') ? 'https://' : 'http://';
```
I copied this from `JsonFormat.php` where it’s used to construct the feed URL. I think this should move to a central location and the base URL (scheme, host, maybe a path prefix) needs to be configurable, e.g. for reverse proxy setups.

somini replied:
Is it possible to have relative links? What does the spec say?

sabberworm replied:
I guess it’s not recommended for feeds as they should be easily transferable…

sabberworm commented on bridges/YoutubeBridge.php:
```diff
@@ -102,9 +136,105 @@ private function ytBridgeAddItem($vid, $title, $author, $desc, $time){
+		// Needs to be a file cache because it needs to be servable by apache
+		$cache = $cacheFac->create('file');
+		$cache->setScope('youtube-dl');
+		$cache->purgeCache(15 * 24 * 60 * 60); // 15 days
```
I think this should be configurable but currently it seems only few bridge behaviours are configurable in the settings and I didn’t want to special-case this.

sabberworm commented on bridges/YoutubeBridge.php:
```diff
@@ -102,9 +136,105 @@ private function ytBridgeAddItem($vid, $title, $author, $desc, $time){
+			'--no-mtime'
+		);
+		if($isAudio) {
+			$args[] = '--extract-audio';
```
I realise some audio formats also work with `--format` but in general, the post processing route was successful for more videos.

sabberworm commented on bridges/YoutubeBridge.php:
```diff
@@ -102,9 +136,105 @@ private function ytBridgeAddItem($vid, $title, $author, $desc, $time){
+		$cacheFac = new CacheFactory();
+		$cacheFac->setWorkingDir(PATH_LIB_CACHES);
+		// Needs to be a file cache because it needs to be servable by apache
+		$cache = $cacheFac->create('file');
```
I know I kinda abuse the caching functionality but I found the features to automatically purge old files and create the cache dir really compelling.

sabberworm commented on bridges/YoutubeBridge.php:
```diff
@@ -102,9 +136,105 @@ private function ytBridgeAddItem($vid, $title, $author, $desc, $time){
+			$returnValue
+		);
+		if($returnValue !== 0) {
+			error_log("Downloading the YouTube video $vid using $command failed: " . implode("\n", $output));
```
Is there a convention of inlining these kinds of error messages into the feed?

somini replied:
The `returnClientError` function. Throws an exception.

https://gitlab.com/somini/rss-bridge/blob/ef0130005cdc150a7ce2dc9215b7f8e7cfb4495d/lib/error.php#L14-34

sabberworm replied:
I’m not sure that’s what I want since it could be only one video in the feed that has a problem. There’s no need to 40x the entire feed.

em92 commented:
Looks like it will drain free Space on instance machine

sabberworm commented on bridges/YoutubeBridge.php:
```diff
@@ -102,9 +136,105 @@ private function ytBridgeAddItem($vid, $title, $author, $desc, $time){
+		$urlHost = (isset($_SERVER['HTTP_HOST'])) ? $_SERVER['HTTP_HOST'] : '';
+
+		// FIXME: Make base URL configurable instead of relying on host header and server vars
+		$url = $urlPrefix . $urlHost . substr($path, strlen($mainDir));
```
I found it a bit odd that the cache dir is exposed under the document root by default but I saw that the PixivBridge also makes use of this already…

sabberworm committed:
chore: fix formatting errors

sabberworm committed:
chore: don’t use empty default value

sabberworm commented:
> Looks like it will drain free Space on instance machine

totally, yes. The entire feature should be opt-in. It kinda is already since you need to install youtube-dl…

Alternatively, we could also introduce a proxy that downloads the video on the fly only when used (decreases storage needs but possibly increases bandwidth and processing time).

somini commented:
This is a crazy contribution, using the cache code to store the data is ingenious. Podcasts from Youtube, amazing. Great job.

Since this requires `youtube-dl` to start filling the disk with videos, it should be safe-ish to merge.

somini commented on bridges/YoutubeBridge.php:
```diff
@@ -102,9 +137,105 @@ private function ytBridgeAddItem($vid, $title, $author, $desc, $time){
+
+		if(empty($attach_format)) {
+			return;
+		}
```
Better check if the `youtube-dl` exe exists here, and fail loudly otherwise. Since the parameters are fixed.

sabberworm replied:
It’s not actually that simple to do a cross-platform check if an executable exists on the $PATH. I’d say simply log whatever error is returned by running youtube-dl and it becomes pretty clear if the error means youtube-dl isn’t installed or on the $PATH.

em92 commented:
There can be `youtube-dl` already installed in host machine, but it does not mean that instance admin is aware that rss-bridge uses YouTube-dl. After update he would get surprise, that he wont like.

Why not using links to invidious instance, from which downloading is enabled?

somini reviewed (commented):
> There can be `youtube-dl` already installed in host machine, but it does not mean that instance admin is aware that rss-bridge uses YouTube-dl. After update he would get surprise, that he wont like.

Hmm, perhaps this should be gated through a config option, disabled by default? This is similar to what we do today with the whitelist.

> Why not using links to invidious instance, from which downloading is enabled?

I tried to check the public invidio.us instance, but download is disabled there, I can't test it.
Does it add the video files as enclosures on the feeds (or better, a direct link to them)? This is the crucial features that allows these feeds to be consumed by podcast apps.

em92 commented:
> I tried to check the public invidio.us instance, but download is disabled there, I can't test it.

You can test on others https://github.com/omarroth/invidious/wiki/Invidious-Instances

> Does it add the video files as enclosures on the feeds (or better, a direct link to them)? This is the crucial features that allows these feeds to be consumed by podcast apps.

Don't know.

somini commented:
> Does it add the video files as enclosures on the feeds (or better, a direct link to them)? This is the crucial features that allows these feeds to be consumed by podcast apps.

Random example from the frontpage: https://invidious.ggc-project.de/feed/channel/UCBa659QWEk1AI4Tg--mrJ2A

It doesn't.

sabberworm committed:
refactor: allow bridges to override getParameters instead of const PARAMETERS

sabberworm committed:
feat(YouTube): make allow_download an explicit opt-in

sabberworm committed:
refactor(YouTube): generate attach_format param definitions from centralized list

sabberworm committed:
feat(YouTube): make cache time for downloads configurable

sabberworm committed:
feat: make base URL configurable

sabberworm committed:
fix(test): use method only after initializing the object under test

sabberworm committed:
chore: fix lint

sabberworm committed:
feat(test): move global require to bootstrap, which also prepares config

sabberworm committed:
fix(YouTube): only add format values to global params if download is allowed

sabberworm commented:
Ok, so I did make the feature opt-in using a config. I also made it so the option doesn’t show up if the feature is disabled (required a bit of refactoring) and I added a central place to configure the base URL, which should simplify reverse-proxy installations. I fixed all the tests, I think, except for the one fixed by #1521.

sabberworm committed:
Merge remote-tracking branch 'origin/master' into youtube-dl

* origin/master: (24 commits)
  [Core] Include Media RSS namespace for Atom feeds
  [GizmodoBridge] Fix bridge (#1538)
  [AirBreizhBridge] Add new bridge (#1544)
  [MozillaBugTrackerBridge] Fix bridge (#1550)
  [UnraidCommunityApplicationsBridge] Add new bridge (#1534)
  [VkBridge] Fix one letter bug on titles (#1555)
  [AuoJMBridge] Follow Website change (#1527)
  [DownDetectorBridge] Fix bridge (#1528)
  [ASRockNewsBridge] Add Bridge (#1526)
  [AwwwardsBridge] New bridge (#1524)
  [MediapartBlogsBridge] Lint
  add nordbayern bridge
  [HeiseBridge] add TechStage support
  [FolhaDeSaoPauloBridge]: Improve URL
  [FolhaDeSaoPauloBridge]: Improve HTML
  Revert feed title
  [MediapartBlogsBridge]: New Bridge
  Change feed title
  Change cache time to 24hr (daily update time)
  Fix title if language not set
  ...

sabberworm commented:
Dear maintainers, let me know if you’re still interested in this. I’m quite happy with https://github.com/frou/yt2pod, so I don’t personally need this anymore.
I still think it’s worth merging, if only for the refactoring I did…
I’ll gladly resolve the conflict, otherwise just close the PR.

em92 commented:
Hi, @sabberworm !
I was against this from the begining. So I will just close PR.
